### DISCUSSION TITLE
How to send a key

### DISCUSSION THREAD

**Discussion by cl0vvn** [FabricMC/fabric-api#1996] [Mod Dev Support] [answered]
I want do a command that will press a key i will send with a command like:

/press w 10 - press w for 10 secs
or
/press esc - press escape

**Comment by warjort** [comment, marked as answer]
Key presses typically don't work that way in any programming system.

Key presses are stored internally by the operating system in a buffer/queue which a program will then poll and respond to by executing some routine. The same routine will likely also be bound to mouse/controller events.
This linkage is usually arbitrary, in Minecraft these could be:
* Player movement controls
* Menu activation
* Block activation
* Item usage
* GUI interaction, eg. button presses
* etc.

So what you need to know is what routine/method is bound to F10 or esc in your example(s) above. 
This will in general be context sensitive based on what UI is displayed.

But assuming you do know what piece of code should be run, 
it will be easier if the routine can be run in the context of the server (this is how most commands work).

Otherwise you need to send a network request to the client which will in turn execute that method when the network message is received.

I suspect you are trying to write an "autoclicker" or some kind of macro automation?
Those are typically implemented external to a program and talk directly to the operating system by pretending to be an input device and/or use an OS specific api. This removes the need to understand the arbitrary ways a program processes inputs.

**Reply by cl0vvn** [reply]
Thanks for the answer!
